File: packages/core/src/ids.ts
/**
 * Branded ID types. They are strings at runtime but distinct at compile time,
 * so you cannot accidentally pass a SandboxId where a SessionId is expected.
 */
export type Brand<T, B extends string> = T & { readonly __brand: B };

export type SessionId = Brand<string, "SessionId">;
export type SandboxId = Brand<string, "SandboxId">;
export type ProviderId = string;

/** A short, URL-safe, sortable-ish id with a human-readable prefix. */
export function newId<T extends string>(prefix: string): Brand<T, string> {
  const rand = cryptoRandom(8);
  return `${prefix}_${rand}` as Brand<T, string>;
}

export const newSessionId = () => newId<SessionId>("ses");
export const newSandboxId = () => newId<SandboxId>("sbx");

function cryptoRandom(bytes: number): string {
  // Node 22 has global crypto.getRandomValues.
  const buf = new Uint8Array(bytes);
  crypto.getRandomValues(buf);
  return Array.from(buf, (b) => b.toString(16).padStart(2, "0")).join("");
}

File: crates/core/src/ids.rs
//! Strongly typed identifiers. They are strings at runtime but distinct types at
//! compile time, so a SandboxId cannot be passed where a SessionId is expected.

use std::fmt;

use serde::{Deserialize, Serialize};
use uuid::Uuid;

macro_rules! string_id {
    ($name:ident, $prefix:literal) => {
        #[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
        pub struct $name(String);

        impl $name {
            /// Generate a fresh id with the type's prefix.
            pub fn new() -> Self {
                let raw = Uuid::new_v4().simple().to_string();
                Self(format!("{}_{}", $prefix, &raw[..12]))
            }

            pub fn as_str(&self) -> &str {
                &self.0
            }
        }

        impl Default for $name {
            fn default() -> Self {
                Self::new()
            }
        }

        impl fmt::Display for $name {
            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
                f.write_str(&self.0)
            }
        }

        impl From<String> for $name {
            fn from(s: String) -> Self {
                Self(s)
            }
        }

        impl From<&str> for $name {
            fn from(s: &str) -> Self {
                Self(s.to_string())
            }
        }
    };
}

string_id!(SessionId, "ses");
string_id!(SandboxId, "sbx");

/// Provider identifier, e.g. "docker" or "e2b".
pub type ProviderId = String;
